homeassistant/helpers/entity.py

```diff
@@ -840,9 +849,47 @@ def _async_write_ha_state(self) -> None:
             return
 
         start = timer()
-        state, attr = self._async_generate_attributes()
+        state, capabilities, attr = self._async_generate_attributes()
         end = timer()
 
+        if entry:
+            # Make sure capabilities in the entity registry are up to date. Capabilities
+            # include capability attributes, device class and supported features
+            device_class: str | None = attr.get(ATTR_DEVICE_CLASS)
```

Comment:
I don't see how, we unconditionally call `hass.states.async_set` and `hass.states.async_set` doesn't indicate if the state was updated or not.

Reply:
Commit 1c52571 makes hass.states.async_set return a bool indicating if the state was updated or not. However, that's not correct because certain entity properties can be overridden by the user meaning gating by state update is not correct; I'm reverting the commit.